Add bowling.rb and test_bowling.rb

igarashi-naoki opened the pull request.

Commit:
Add bowling.rb and test_bowling.rb

Commit:
fix complexity Warning of rubocop

複雑度についてrubocopに警告された。
ChatGPTに聞いて、メソッドとして切り出すことで複雑度を下げ、rubocopをパスした。

Commit:
change to display in stdout

igarashi-naoki committed:
標準出力を参照してテストするように変更

yoshitsugu commented on 03.bowling/test_bowling.rb:
```diff
@@ -0,0 +1,15 @@
+require 'minitest/autorun'
+require_relative 'bowling'
+
+class BowlingTest < Minitest::Test
```
テストがあるのはよいですね :+1:

yoshitsugu commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,58 @@
+      shots << s.to_i
+    end
+  end
+  frames = []
+  shots.each_slice(2) { |s| frames << s }
+  frames
```
豆知識程度ですが、 `each_slice` はブロックなしで Enumeratorを返しますので、 `to_a` すれば配列にできます。

https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/each_slice.html
https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/entries.html

yoshitsugu commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,58 @@
+  frames = shaping_to_frame(score)
+  point = 0
+  frames.each_with_index do |frame, idx|
+    break if idx == 9 # skip 10th frame
```
結局10フレーム目とストライクやスペアじゃないパターンは `frame.sum` のみが加算され、ストライやスペアの場合は `frame.sum` とボーナススコアが加算される、ということですよね。

```ruby
 frames.each_with_index do |frame. idx|
  point += frame.sum
  next if idx >= 9 || frame.sum != 10

  # これ以降は9フレーム以内、かつ、スペアかストライクのどちらか
  if frame[0] == 10
    # ストライクのボーナススコア
  else
    # スペアのボーナススコア
  end
end
```
のようにしてあげるとちょっとスッキリするのかなと思いました。

yoshitsugu reviewed (commented):
コメントしました！

igarashi-naoki committed:
レビュー内容の反映

1. スコア計算時の分岐をより簡潔に変更
「現フレームの点数加算」という処理は共通で実装し、それ以降フレームの種類によってボーナススコアを加算するように変更

2. 入力文字列をフレーム変換するメソッドのリファクタリング

yoshitsugu reviewed (approved):
OKです〜
